Feat(trends):t_score function

wmak opened the pull request:
- This adds a [t score](https://en.wikipedia.org/wiki/Welch%27s_t-test) function to help us identify significant trends
- This introduces variance as a range function as well, to support the calculation of the t score
- Adding a default t_score filter, we'll eventually have the ability for the user to change the number as "low/med/high" but for now filtering >1 to remove the ones we have no confidence in.

wmak committed:
fix: Non avg displays show percentage instead of NaN

- Because avg has to be included to calculate the t_score, we can't just
  iterate on all the functions like this anymore. Now only checking the
  trendFunction instead

wmak committed:
feat(trends): Add a t score function

- This adds a t score function to help us identify significant trends
- This introduces variance as a range function as well, to support the
  calculation of the t score

wmak committed:
feat(trends): Add the ability to filter on t_score

- Adding a default t_score filter, we'll eventually have the ability for
  the user to change the number as "low/med/high" but for now filtering
  >1 to remove the ones we have no confidence in.

wmak committed:
fix: Missed a debug

wmak committed:
fix: acceptance tests

- Also refactor them a bit so its a bit cleaner/pythonic

wmak commented on src/sentry/api/endpoints/organization_events_trends.py:
```diff
@@ -39,13 +40,39 @@ class OrganizationEventsTrendsEndpointBase(OrganizationEventsV2EndpointBase):
             "alias": "user_misery_range_",
         },
+        "variance": {
+            "format": "variance_range(transaction.duration, {start}, {end}, {index})",
```
👍 I think eventually all of these will need to be updated to support measurements.

trillville commented:
Oh shoot - one thing I forgot to mention in our meeting today: t-score calculations are very sensitive to outliers, would it make sense/be feasible to be a bit more aggressive with the default transaction maximum length filters?

trillville commented on src/sentry/static/sentry/app/views/performance/trends/utils.tsx:
```diff
@@ -381,14 +379,10 @@ export function movingAverage(data, index, size) {
+  let limitQuery = 'trend_percentage():<1 t_score():>1';
   if (trendChangeType === TrendChangeType.REGRESSION) {
-    limitQuery = aliasedPercentage + ':>1';
+    limitQuery = 'trend_percentage():>1 t_score():<-1';
```
would personally vote for +/- 2 as the default cutoff, +/- 1 corresponds to a super low 70% confidence level (http://uregina.ca/~gingrich/tt.pdf)

wmak replied:
:+1: There's an upcoming PR to replace this with the Low/Medium/High mentioned

wmak commented:
:+1: I'll update the `transaction.duration:<=60min` to `transaction.duration:<=15min`

wmak committed:
Merge branch 'master' into feat/tscore-function

wmak committed:
ref: Updating max duration to 15minutes

- Also moving this to a variable to make it easier to update

markstory commented on tests/snuba/api/endpoints/test_organization_events_trends.py:
```diff
@@ -537,6 +563,8 @@ def test_misery_trend_function(self):
             "percentage_user_misery_range_2_user_misery_range_1": 2.0,
         }
+        for key, value in expected_data.items():
+            assert events["data"][0][key] == value
```
You might want to consider a helper for this like `assert_event(events["data"][0], expected_data)`

wmak committed:
ref: Moving assert into a helper function

- Reduced some repeat with the expected_data dict
- Added key to the assert_event so if a test fails its clearer which key
  it failed on
